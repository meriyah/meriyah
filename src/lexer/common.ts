import { Chars } from '../chars';
import { Flags } from '../common';
import { type Parser } from '../parser/parser';
import { Token } from '../token';

export const enum LexerState {
  None = 0,
  NewLine = 1 << 0,
  LastIsCR = 1 << 2,
}

export const enum NumberKind {
  ImplicitOctal = 1 << 0,
  Binary = 1 << 1,
  Octal = 1 << 2,
  Hex = 1 << 3,
  Decimal = 1 << 4,
  NonOctalDecimal = 1 << 5,
  Float = 1 << 6,
  ValidBigIntKind = 1 << 7,
  DecimalNumberKind = Decimal | NonOctalDecimal,
}

/**
 * Advances this lexer's current index.
 *
 * @param parser The parser instance
 */
export function advanceChar(parser: Parser): number {
  parser.column++;
  return (parser.currentChar = parser.source.charCodeAt(++parser.index));
}

/**
 * Consumes possible surrogate pair, return the merged code point or 0
 *
 * @param parser The parser instance
 */
export function consumePossibleSurrogatePair(parser: Parser): number {
  // See: https://tc39.github.io/ecma262/#sec-ecmascript-language-types-string-type
  const hi = parser.currentChar;
  // Not surrogate
  if ((hi & 0xfc00) !== Chars.LeadSurrogateMin) return Chars.Null;
  const lo = parser.source.charCodeAt(parser.index + 1);
  // Unpaired surrogate
  if ((lo & 0xfc00) !== Chars.TrailSurrogateMin) return Chars.Null;
  // Merged code point
  return Chars.NonBMPMin + ((hi & 0x3ff) << 10) + (lo & 0x3ff);
}

/**
 * Use to consume a line feed instead of `scanNewLine`.
 */
export function consumeLineFeed(parser: Parser, state: LexerState): void {
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.flags |= Flags.NewLine;
  if ((state & LexerState.LastIsCR) === 0) {
    parser.column = 0;
    parser.line++;
  }
}

export function scanNewLine(parser: Parser): void {
  parser.flags |= Flags.NewLine;
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.column = 0;
  parser.line++;
}

// ECMA-262 11.2 White Space
// gC=Zs, U+0009, U+000B, U+000C, U+FEFF
export function isExoticECMAScriptWhitespace(ch: number): boolean {
  return (
    ch === Chars.NonBreakingSpace ||
    ch === Chars.ZeroWidthNoBreakSpace ||
    ch === Chars.NextLine ||
    ch === Chars.Ogham ||
    (ch >= Chars.EnQuad && ch <= Chars.ZeroWidthSpace) ||
    ch === Chars.NarrowNoBreakSpace ||
    ch === Chars.MathematicalSpace ||
    ch === Chars.IdeographicSpace ||
    ch === Chars.ThinSpace ||
    ch === Chars.ByteOrderMark
  );
}

/**
 * Converts a value to a hex value
 *
 * @param code CodePoint
 */
export function toHex(code: number): number {
  return code < Chars.UpperA ? code - Chars.Zero : (code - Chars.UpperA + 10) & 0xf;
}

/**
 * Converts a token to a string representation
 *
 * @param t Token
 */
export function convertTokenType(t: Token): string {
  switch (t) {
    case Token.NumericLiteral:
      return 'NumericLiteral';
    case Token.StringLiteral:
      return 'StringLiteral';
    case Token.FalseKeyword:
    case Token.TrueKeyword:
      return 'BooleanLiteral';
    case Token.NullKeyword:
      return 'NullLiteral';
    case Token.RegularExpression:
      return 'RegularExpression';
    case Token.TemplateContinuation:
    case Token.TemplateSpan:
    case Token.Template:
      return 'TemplateLiteral';
    default:
      if ((t & Token.IsIdentifier) === Token.IsIdentifier) return 'Identifier';
      if ((t & Token.Keyword) === Token.Keyword) return 'Keyword';

      return 'Punctuator';
  }
}
