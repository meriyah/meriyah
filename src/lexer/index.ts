export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
