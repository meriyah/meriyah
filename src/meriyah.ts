import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
