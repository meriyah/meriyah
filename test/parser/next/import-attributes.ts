import { Context } from '../../../src/common';
import { pass, fail } from '../../test-utils';
import * as t from 'assert';
import { parseSource } from '../../../src/parser';

describe('Next - Import Attributes', () => {
  for (const arg of [
    `import 'bar' with { type: 'what' };`,
    `import foo from 'bar' with { type: 'json' };`,
    `import foo from 'bar' with { type: 'json', 'data-type': 'json' };`,
    `import foo from 'bar' with { "type": 'json' };`,
    `import foo from 'bar' with { "type": 'json', 'data-type': 'json' };`,
    `import {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { type: 'json' };`,
    `import {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { "type": 'json' };`,
    `import * as ns from './json-via-namespace_FIXTURE.json' with { type: 'json' };`,
    `import * as ns from './json-via-namespace_FIXTURE.json' with { "type": 'json' };`,
    `import { random } from './random.ts' with { type: 'macro' };`,
    `import { random } from './random.ts' with { "type": 'macro' };`,
    `import { random as foo } from './random.ts' with { type: 'macro' };`,
    `import { "random" as foo } from './random.ts' with { type: 'macro' };`,
    `import('module', { type: 'json' });`,
    `import('module', { lazy: true });`,
    `import('module', { dynamic: false });`,
    `import('module', { 'data-type': 'json' });`,
    `import('module', { 'data-version': '1.0' });`,
    `import('module', { type: 'text/javascript', version: '2.0' });`,
    `import('module', { 'type': 'module', 'import-attr': true });`,
    `'use strict'; import('module', { type: 'json' });`,
    `async function load() { return import('module', { type: 'json' }); }`,
    `(() => import('module', { type: 'json' }))()`,
    `let dynamicImport = import('module', { type: 'json' });`,
    `let dynamicImport = import('module', { "type": 'json' });`,
    `const importWithAttributes = import('module', { type: 'json' });`,
    `({ async load() { return import('module', { type: 'json' }); } })`,
    `({ async load() { return import('module', { "type": 'json' }); } })`,
    `function* gen() { yield import('module', { type: 'json' }); }`,
    `function* gen() { yield import('module', { "type": 'json' }); }`,
    `for await (let module of [import('module', { type: 'json' })]) {}`,
    `for await (let module of [import('module', { "type": 'json' })]) {}`,
    `import('module', { type: 'json' }).then(module => { /* ... */ });`,
    `import('module', { 'data-type': 'json' }).then(module => { /* ... */ });`,
    `const result = await import('module', { type: 'json' });`,
    `const result = await import('module', { "type": 'json' });`,
    `import x from './import-attribute-1_FIXTURE.js' with {};`,
    `import './import-attribute-2_FIXTURE.js' with {};`,
    `export * from './import-attribute-3_FIXTURE.js' with {};`,
    `(async function () { return import('./2nd-param_FIXTURE.js', await undefined);}())`,
    `var iter = function*() { beforeCount += 1, import('', yield), afterCount += 1;}();`,
    `var promise; for (promise = import('./2nd-param_FIXTURE.js', 'test262' in {} || undefined); false; );`,
    `export {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { type: 'json' };`,
    `export {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { "type": 'json' };`,
    `export * as foo from './foo.json' with { "type": 'json' };`,
    // TODO: to follow up with spec
    // Current JSON modules spec didn't prevent following line.
    `export * from './foo.json' with { "type": 'json' };`,
    `export { random } from './random.ts' with { type: 'macro' };`,
    `export { random } from './random.ts' with { "type": 'macro' };`
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, undefined, Context.OptionsNext | Context.Strict | Context.Module);
      });
    });
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(
          `${arg}`,
          undefined,
          Context.OptionsNext | Context.OptionsWebCompat | Context.Strict | Context.Module
        );
      });
    });
  }

  fail('Expressions - Import Attributes (fail)', [
    ['import("module", { type: "json" }, "extra")', Context.OptionsNext],
    ['import("module", { type: "json" }, "extra")', Context.None],
    ['import("module", { type: "json", "extra": })', Context.OptionsNext | Context.Strict | Context.Module],
    ['import("module", ...extra)', Context.OptionsNext | Context.Strict | Context.Module],
    ['import("module", { type: "json", "extra": "value" ', Context.OptionsNext | Context.Strict | Context.Module],
    [
      'import("module", { type: "json", "extra": "value" }, "another")',
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    ['import("module", { type: "json" }, "extra")', Context.OptionsNext | Context.Strict | Context.Module],
    [
      'import foo from "bar" with { type: "json", "data-type": "json"',
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    ['import foo from "bar" with { type: ', Context.OptionsNext | Context.Strict | Context.Module],
    ['import foo from "bar" with { type: "json", ', Context.OptionsNext | Context.Strict | Context.Module],
    ['import foo from "bar" with { type: "json", "data-type": ', Context.OptionsNext | Context.Strict | Context.Module],
    [
      'import foo from "bar" with { type: "json", "data-type": "json" ',
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    [
      `import {name} from './json-named-bindings_FIXTURE.json' with { type: 'json' };`,
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    [
      `import {name} from './json-named-bindings_FIXTURE.json' with { "type": 'json' };`,
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    [
      `import x from './import-attribute-1_FIXTURE.js' with {
      type: 'json',
      'typ\u0065': ''
    };`,
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    [
      'import { default as a, foo } from "./foo.json" with { type: "json" };',
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    [
      'import a, { foo } from "./foo.json" with { type: "json" };',
      Context.OptionsNext | Context.Strict | Context.Module
    ],
    [`import 'bar' with { type: 'json' };`, Context.OptionsNext | Context.Module],
    ['import foo from "bar" with { 1: "foo" };', Context.OptionsNext | Context.Module],
    ['import foo from "bar" with { type: 1 };', Context.OptionsNext | Context.Module],
    ['import foo from "bar" with { type: [1] };', Context.OptionsNext | Context.Module],
    ['import foo from "bar" with { type: null };', Context.OptionsNext | Context.Module],
    ['import foo from "bar" with { type: undefined };', Context.OptionsNext | Context.Module],
    ['import foo from "bar" with { type: "json", foo: {} };', Context.OptionsNext | Context.Module],
    [`export { foo } from './foo.json' with { type: 'json' };`, Context.OptionsNext | Context.Module],
    [`export foo, { foo2 } from './foo.json' with { "type": 'json' };`, Context.OptionsNext | Context.Module]
  ]);

  pass('Next - Import Attributes (pass)', [
    [
      `import('module', { type: 'json' });`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'ImportExpression',
              source: {
                type: 'Literal',
                value: 'module'
              },
              options: {
                type: 'ObjectExpression',
                properties: [
                  {
                    computed: false,
                    type: 'Property',
                    key: {
                      type: 'Identifier',
                      name: 'type'
                    },
                    value: {
                      type: 'Literal',
                      value: 'json'
                    },
                    kind: 'init',
                    method: false,
                    shorthand: false
                  }
                ]
              }
            }
          }
        ]
      }
    ],
    [
      `import('module', { 'data-type': 'json' });`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'ImportExpression',
              source: {
                type: 'Literal',
                value: 'module'
              },
              options: {
                type: 'ObjectExpression',
                properties: [
                  {
                    computed: false,
                    type: 'Property',
                    key: {
                      type: 'Literal',
                      value: 'data-type'
                    },
                    value: {
                      type: 'Literal',
                      value: 'json'
                    },
                    kind: 'init',
                    method: false,
                    shorthand: false
                  }
                ]
              }
            }
          }
        ]
      }
    ],
    [
      `async function load() { return import('module', { type: 'json' }); }`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        body: [
          {
            async: true,
            body: {
              body: [
                {
                  argument: {
                    options: {
                      properties: [
                        {
                          computed: false,
                          key: {
                            name: 'type',
                            type: 'Identifier'
                          },
                          kind: 'init',
                          method: false,
                          shorthand: false,
                          type: 'Property',
                          value: {
                            type: 'Literal',
                            value: 'json'
                          }
                        }
                      ],
                      type: 'ObjectExpression'
                    },
                    source: {
                      type: 'Literal',
                      value: 'module'
                    },
                    type: 'ImportExpression'
                  },
                  type: 'ReturnStatement'
                }
              ],
              type: 'BlockStatement'
            },
            generator: false,
            id: {
              name: 'load',
              type: 'Identifier'
            },
            params: [],
            type: 'FunctionDeclaration'
          }
        ],
        sourceType: 'module',
        type: 'Program'
      }
    ],
    [
      `for await (let module of [import('module', { type: 'json' })]) {}`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ForOfStatement',
            left: {
              type: 'VariableDeclaration',
              declarations: [
                {
                  type: 'VariableDeclarator',
                  id: {
                    type: 'Identifier',
                    name: 'module'
                  },
                  init: null
                }
              ],
              kind: 'let'
            },
            right: {
              type: 'ArrayExpression',
              elements: [
                {
                  type: 'ImportExpression',
                  source: {
                    type: 'Literal',
                    value: 'module'
                  },
                  options: {
                    type: 'ObjectExpression',
                    properties: [
                      {
                        computed: false,
                        type: 'Property',
                        key: {
                          type: 'Identifier',
                          name: 'type'
                        },
                        value: {
                          type: 'Literal',
                          value: 'json'
                        },
                        kind: 'init',
                        method: false,
                        shorthand: false
                      }
                    ]
                  }
                }
              ]
            },
            await: true,
            body: {
              type: 'BlockStatement',
              body: []
            }
          }
        ]
      }
    ],
    [
      'import foo from "bar" with { type: "json" };',
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        body: [
          {
            source: {
              type: 'Literal',
              value: 'bar'
            },
            specifiers: [
              {
                local: {
                  name: 'foo',
                  type: 'Identifier'
                },
                type: 'ImportDefaultSpecifier'
              }
            ],
            type: 'ImportDeclaration',
            attributes: [
              {
                key: {
                  name: 'type',
                  type: 'Identifier'
                },
                value: {
                  type: 'Literal',
                  value: 'json'
                },
                type: 'ImportAttribute'
              }
            ]
          }
        ],
        sourceType: 'module',
        type: 'Program'
      }
    ],
    [
      'import foo from "bar" with { type: "json", "data-type": "json" };',
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        body: [
          {
            source: {
              type: 'Literal',
              value: 'bar'
            },
            specifiers: [
              {
                local: {
                  name: 'foo',
                  type: 'Identifier'
                },
                type: 'ImportDefaultSpecifier'
              }
            ],
            type: 'ImportDeclaration',
            attributes: [
              {
                key: {
                  name: 'type',
                  type: 'Identifier'
                },
                value: {
                  type: 'Literal',
                  value: 'json'
                },
                type: 'ImportAttribute'
              },
              {
                key: {
                  type: 'Literal',
                  value: 'data-type'
                },
                value: {
                  type: 'Literal',
                  value: 'json'
                },
                type: 'ImportAttribute'
              }
            ]
          }
        ],
        sourceType: 'module',
        type: 'Program'
      }
    ],
    [
      `var promise; for (promise = import('./2nd-param_FIXTURE.js', 'test262' in {} || undefined); false; );`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        body: [
          {
            declarations: [
              {
                id: {
                  name: 'promise',
                  type: 'Identifier'
                },
                init: null,
                type: 'VariableDeclarator'
              }
            ],
            kind: 'var',
            type: 'VariableDeclaration'
          },
          {
            body: {
              type: 'EmptyStatement'
            },
            init: {
              left: {
                name: 'promise',
                type: 'Identifier'
              },
              operator: '=',
              right: {
                options: {
                  left: {
                    left: {
                      type: 'Literal',
                      value: 'test262'
                    },
                    operator: 'in',
                    right: {
                      properties: [],
                      type: 'ObjectExpression'
                    },
                    type: 'BinaryExpression'
                  },
                  operator: '||',
                  right: {
                    name: 'undefined',
                    type: 'Identifier'
                  },
                  type: 'LogicalExpression'
                },
                source: {
                  type: 'Literal',
                  value: './2nd-param_FIXTURE.js'
                },
                type: 'ImportExpression'
              },
              type: 'AssignmentExpression'
            },
            test: {
              type: 'Literal',
              value: false
            },
            type: 'ForStatement',
            update: null
          }
        ],
        sourceType: 'module',
        type: 'Program'
      }
    ],
    [
      `export * from './foo' with { type: 'json' }`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExportAllDeclaration',
            source: {
              type: 'Literal',
              value: './foo'
            },
            exported: null,
            attributes: [
              {
                type: 'ImportAttribute',
                key: {
                  type: 'Identifier',
                  name: 'type'
                },
                value: {
                  type: 'Literal',
                  value: 'json'
                }
              }
            ]
          }
        ]
      }
    ],
    [
      `export * as foo from './foo' with { type: 'json' };`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExportAllDeclaration',
            source: {
              type: 'Literal',
              value: './foo'
            },
            exported: {
              type: 'Identifier',
              name: 'foo'
            },
            attributes: [
              {
                type: 'ImportAttribute',
                key: {
                  type: 'Identifier',
                  name: 'type'
                },
                value: {
                  type: 'Literal',
                  value: 'json'
                }
              }
            ]
          }
        ]
      }
    ],
    [
      `export {} from './foo' with { type: 'html' };`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: null,
            specifiers: [],
            source: {
              type: 'Literal',
              value: './foo'
            },
            attributes: [
              {
                type: 'ImportAttribute',
                key: {
                  type: 'Identifier',
                  name: 'type'
                },
                value: {
                  type: 'Literal',
                  value: 'html'
                }
              }
            ]
          }
        ]
      }
    ],
    [
      `export { foo } from './foo' with { type: 'html' }`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: null,
            specifiers: [
              {
                type: 'ExportSpecifier',
                local: {
                  type: 'Identifier',
                  name: 'foo'
                },
                exported: {
                  type: 'Identifier',
                  name: 'foo'
                }
              }
            ],
            source: {
              type: 'Literal',
              value: './foo'
            },
            attributes: [
              {
                type: 'ImportAttribute',
                key: {
                  type: 'Identifier',
                  name: 'type'
                },
                value: {
                  type: 'Literal',
                  value: 'html'
                }
              }
            ]
          }
        ]
      }
    ],
    [
      `export { foo, } from './foo' with { type: 'html' };`,
      Context.Module | Context.Strict | Context.OptionsNext,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: null,
            specifiers: [
              {
                type: 'ExportSpecifier',
                local: {
                  type: 'Identifier',
                  name: 'foo'
                },
                exported: {
                  type: 'Identifier',
                  name: 'foo'
                }
              }
            ],
            source: {
              type: 'Literal',
              value: './foo'
            },
            attributes: [
              {
                type: 'ImportAttribute',
                key: {
                  type: 'Identifier',
                  name: 'type'
                },
                value: {
                  type: 'Literal',
                  value: 'html'
                }
              }
            ]
          }
        ]
      }
    ]
  ]);
});
