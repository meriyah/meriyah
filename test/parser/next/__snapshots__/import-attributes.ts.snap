// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Next - Import Attributes > Next - Import Attributes (pass) > async function load() { return import('module', { type: 'json' }); } 1`] = `
{
  "body": [
    {
      "async": true,
      "body": {
        "body": [
          {
            "argument": {
              "options": {
                "properties": [
                  {
                    "computed": false,
                    "key": {
                      "name": "type",
                      "type": "Identifier",
                    },
                    "kind": "init",
                    "method": false,
                    "shorthand": false,
                    "type": "Property",
                    "value": {
                      "type": "Literal",
                      "value": "json",
                    },
                  },
                ],
                "type": "ObjectExpression",
              },
              "source": {
                "type": "Literal",
                "value": "module",
              },
              "type": "ImportExpression",
            },
            "type": "ReturnStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "load",
        "type": "Identifier",
      },
      "params": [],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > export * as foo from './foo' with { type: 'json' }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "exported": {
        "name": "foo",
        "type": "Identifier",
      },
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "type": "ExportAllDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > export * from './foo' with { type: 'json' } 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "exported": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "type": "ExportAllDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > export { foo } from './foo' with { type: 'html' } 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "html",
          },
        },
      ],
      "declaration": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "specifiers": [
        {
          "exported": {
            "name": "foo",
            "type": "Identifier",
          },
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ExportSpecifier",
        },
      ],
      "type": "ExportNamedDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > export { foo, } from './foo' with { type: 'html' }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "html",
          },
        },
      ],
      "declaration": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "specifiers": [
        {
          "exported": {
            "name": "foo",
            "type": "Identifier",
          },
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ExportSpecifier",
        },
      ],
      "type": "ExportNamedDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > export {} from './foo' with { type: 'html' }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "html",
          },
        },
      ],
      "declaration": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "specifiers": [],
      "type": "ExportNamedDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > for await (let module of [import('module', { type: 'json' })]) {} 1`] = `
{
  "body": [
    {
      "await": true,
      "body": {
        "body": [],
        "type": "BlockStatement",
      },
      "left": {
        "declarations": [
          {
            "id": {
              "name": "module",
              "type": "Identifier",
            },
            "init": null,
            "type": "VariableDeclarator",
          },
        ],
        "kind": "let",
        "type": "VariableDeclaration",
      },
      "right": {
        "elements": [
          {
            "options": {
              "properties": [
                {
                  "computed": false,
                  "key": {
                    "name": "type",
                    "type": "Identifier",
                  },
                  "kind": "init",
                  "method": false,
                  "shorthand": false,
                  "type": "Property",
                  "value": {
                    "type": "Literal",
                    "value": "json",
                  },
                },
              ],
              "type": "ObjectExpression",
            },
            "source": {
              "type": "Literal",
              "value": "module",
            },
            "type": "ImportExpression",
          },
        ],
        "type": "ArrayExpression",
      },
      "type": "ForOfStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > import foo from "bar" with { type: "json" }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "source": {
        "type": "Literal",
        "value": "bar",
      },
      "specifiers": [
        {
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ImportDefaultSpecifier",
        },
      ],
      "type": "ImportDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > import foo from "bar" with { type: "json", "data-type": "json" }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
        {
          "key": {
            "type": "Literal",
            "value": "data-type",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "source": {
        "type": "Literal",
        "value": "bar",
      },
      "specifiers": [
        {
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ImportDefaultSpecifier",
        },
      ],
      "type": "ImportDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > import('module', { 'data-type': 'json' }); 1`] = `
{
  "body": [
    {
      "expression": {
        "options": {
          "properties": [
            {
              "computed": false,
              "key": {
                "type": "Literal",
                "value": "data-type",
              },
              "kind": "init",
              "method": false,
              "shorthand": false,
              "type": "Property",
              "value": {
                "type": "Literal",
                "value": "json",
              },
            },
          ],
          "type": "ObjectExpression",
        },
        "source": {
          "type": "Literal",
          "value": "module",
        },
        "type": "ImportExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > import('module', { type: 'json' }); 1`] = `
{
  "body": [
    {
      "expression": {
        "options": {
          "properties": [
            {
              "computed": false,
              "key": {
                "name": "type",
                "type": "Identifier",
              },
              "kind": "init",
              "method": false,
              "shorthand": false,
              "type": "Property",
              "value": {
                "type": "Literal",
                "value": "json",
              },
            },
          ],
          "type": "ObjectExpression",
        },
        "source": {
          "type": "Literal",
          "value": "module",
        },
        "type": "ImportExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Next - Import Attributes (pass) > var promise; for (promise = import('./2nd-param_FIXTURE.js', 'test262' in {} || undefined); false; ); 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "promise",
            "type": "Identifier",
          },
          "init": null,
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
    {
      "body": {
        "type": "EmptyStatement",
      },
      "init": {
        "left": {
          "name": "promise",
          "type": "Identifier",
        },
        "operator": "=",
        "right": {
          "options": {
            "left": {
              "left": {
                "type": "Literal",
                "value": "test262",
              },
              "operator": "in",
              "right": {
                "properties": [],
                "type": "ObjectExpression",
              },
              "type": "BinaryExpression",
            },
            "operator": "||",
            "right": {
              "name": "undefined",
              "type": "Identifier",
            },
            "type": "LogicalExpression",
          },
          "source": {
            "type": "Literal",
            "value": "./2nd-param_FIXTURE.js",
          },
          "type": "ImportExpression",
        },
        "type": "AssignmentExpression",
      },
      "test": {
        "type": "Literal",
        "value": false,
      },
      "type": "ForStatement",
      "update": null,
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;
